### PR TITLE
[CALCITE-5061] Improve recursive application of the field trimming

### PR DESCRIPTION
**Pull request by rubenada**
[CALCITE-5061](https://issues.apache.org/jira/browse/CALCITE-5061) - Improve recursive application of the field trimming: if a RelNode itself cannot be trimmed, try to trim inside its subtree as much as possible

### CODE REVIEW

**Review by hsyuan — APPROVED** [review]
LGTM.

File: core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java [hunk 1]
```diff
@@ -352,11 +352,24 @@ public TrimResult trimFields(
...
+    }
+
+    // We don't know how to trim this RelNode, but we can try to trim inside its inputs
+    List<RelNode> newInputs = new ArrayList<>(rel.getInputs().size());
+    for (RelNode input : rel.getInputs()) {
+      ImmutableBitSet inputFieldsUsed = ImmutableBitSet.range(input.getRowType().getFieldCount());
+      TrimResult trimResult = dispatchTrimFields(input, inputFieldsUsed, extraFields);
+      if (!trimResult.right.isIdentity()) {
+        throw new IllegalArgumentException();
+      }
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:368]: > throw new IllegalArgumentException();

Should we add some clear exception messages instead?

**Review by chunweilei — APPROVED** [review]
LGTM. Left a minor comment.

File: core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java (hunk 1 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:369]: Exception message improved.

**Comment by rubenada** [comment]
Thanks for the review @hsyuan @chunweilei .
I have squashed commits into a single one. I will merge this PR in the coming days if no objection appears.

**Pull request merged**